src/data/polynomial/laurent.lean

```diff
@@ -0,0 +1,186 @@
+
+## Implementation notes
+
+We define Laurent polynomials as `add_monoid_algebra R ℤ`.
+Thus, they are essentially `finsupp`s `ℤ →₀ R`.
+This choice differs from the current irreducible design of `polynomial`, that instead shields away
+the implementation via `finsupp`s.  It is closer to the original definition of polynomials.
+
+As a consequence, `laurent_polynomial` plays well with polynomials, but there is a little roughness
+in establishing the API, since the `finsupp` implementation of `polynomial R` is well-shielded.
+
+Unlike the case of polynomials, I felt that the exponent notation was not too easy to use, as only
+natural exponents would be allowed.  Moreover, in the end, it seems likely that we should aim to
+perform computations on exponents in `ℤ` anyway and separating this via the symbol `T` seems
+convenient.
+
+I made a *heavy* use of `simp` lemmas, aiming to bring Laurent polynomials to the form `C a * T n`.
+Any comments or suggestions for improvements is greatly appreciated!
+
+##  Future work
+Lots is missing!  I would certainly like to show that `R[T;T⁻¹]` is the localization of `R[X]`
+inverting `X`.  This should be mostly in place, given `exists_T_pow` (which is part of PR #13415).
+-/
+
+open_locale polynomial big_operators
+open polynomial add_monoid_algebra finsupp
+noncomputable theory
+
+variables {R : Type*}
+
+/--  The semiring of Laurent polynomials with coefficients in the semiring `R`.
+We denote it by `R[T;T⁻¹]`.
+The ring homomorphism `C : R →+* R[T;T⁻¹]` includes `R` as the constant polynomials. -/
+abbreviation laurent_polynomial (R : Type*) [semiring R] := add_monoid_algebra R ℤ
+
+local notation R`[T;T⁻¹]`:9000 := laurent_polynomial R
+
+/--  The ring homomorphism, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent [semiring R] : R[X] →+* R[T;T⁻¹] :=
+(map_domain_ring_hom R int.of_nat_hom).comp (to_finsupp_iso R)
+
+/-- This is not a simp lemma, as it is usually preferable to use the lemmas about `C` and `X`
+instead. -/
+lemma polynomial.to_laurent_apply [semiring R] (p : R[X]) :
+  p.to_laurent = p.to_finsupp.map_domain coe := rfl
+
+/--  The `R`-algebra map, taking a polynomial with coefficients in `R` to a Laurent polynomial
+with coefficients in `R`. -/
+def polynomial.to_laurent_alg [comm_semiring R] :
+  R[X] →ₐ[R] R[T;T⁻¹] :=
+begin
+  refine alg_hom.comp _ (to_finsupp_iso_alg R).to_alg_hom,
+  exact (map_domain_alg_hom R R int.of_nat_hom),
+end
+
+@[simp]
+lemma polynomial.to_laurent_alg_apply [comm_semiring R] (f : R[X]) :
+  f.to_laurent_alg = f.to_laurent := rfl
+
+namespace laurent_polynomial
+
+section semiring
+variables [semiring R]
+
+lemma single_zero_one_eq_one : (single 0 1 : R[T;T⁻¹]) = (1 : R[T;T⁻¹]) := rfl
+
+/-!  ### The functions `C` and `T`. -/
+/--  The ring homomorphism `C`, including `R` into the ring of Laurent polynomials over `R` as
+the constant Laurent polynomials. -/
+def C : R →+* R[T;T⁻¹] :=
```

Comment:
Can you add this as an algebra map?

Reply:
Doesn't that already exist as `algebra.of_id`?

Reply:
I'm not at a computer, but on line 78 there should also be something relevant.  Note that here I do not think that `R` is commutative, so the algebra map should be somewhere else anyway.

Reply:
You're right, this is `algebra.of_id`. Maybe a `rfl` lemma `C_apply`?

Reply:
Riccardo, I am not sure what lemma you have in mind.  Is it something like `C a = a \smul 1`?

Reply:
Just that `C r = algebra.of_id r`.

Reply:
Do we have that for `polynomial`? I think we just want that `C = algebra_map _ _` (or the applied version; whatever is consistent with `polynomial`), which in turn is defeq to `algebra.of_id`.

Reply:
Ok, I'll look for the polynomial one and will produce the analogous for Laurent!

Reply:
I think that the lemma that you had in mind was `algebra_map_apply` and I added it.  It might move to a different part of the file, once the stuff with `comm_semiring` will appear in later PRs.

Reply:
For polynomials we have [polynomial.C_eq_algebra_map](https://leanprover-community.github.io/mathlib_docs/data/polynomial/algebra_map.html#polynomial.C_eq_algebra_map).